modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,433 @@
+
+  if (navigator.appVersion.indexOf('Win') !== -1) {
+    envParams.osName = 'Windows';
+  } else if (navigator.appVersion.indexOf('Mac') !== -1) {
+    envParams.osName = 'MacOS';
+  } else if (navigator.appVersion.indexOf('X11') !== -1) {
+    envParams.osName = 'Unix';
+  } else if (navigator.appVersion.indexOf('Linux') !== -1) {
+    envParams.osName = 'Linux';
+  }
+
+  let browserLanguage = navigator.language || navigator.userLanguage;
+  let acceptLanguage = browserLanguage.replace('_', '-');
+
+  envParams.language = acceptLanguage;
+
+  utils.logInfo('Domain -> ', envParams.domain);
+  utils.logInfo('Page -> ', envParams.page);
+  utils.logInfo('Lang -> ', envParams.lang);
+  utils.logInfo('OS -> ', envParams.osName);
+  utils.logInfo('User Agent -> ', envParams.userAgent);
+  utils.logInfo('Primary Language -> ', envParams.language);
+}
+
+export const imps = new Map();
+
+manageEnvParams();
+
+function handleValidRTB2Dot4(bid) {
+  const bannerInfo = bid.mediaTypes?.banner;
+  const nativeInfo = bid.mediaTypes?.native;
+  const videoInfo = bid.mediaTypes?.video;
+  const isValid = !!(
+    bid.params.endpoint &&
+    bid.params.userIp &&
+    bid.params.hasOwnProperty('userId') &&
+    bid.params.zoneId &&
+    bid.params.partner &&
+    bid.params.fid &&
+    bid.params.siteId &&
+    bid.params.impressionId &&
+    bid.params.country &&
+    bid.params.country.length > 0 &&
+    (bannerInfo || nativeInfo || videoInfo) &&
+    !!(bid.params.bidfloor && bid.params.bidfloorcur) &&
+    (nativeInfo ? bid.params.native &&
+      nativeInfo.ortb.assets &&
+      nativeInfo.ortb.assets.some(asset => !!asset.img) : true) &&
+    (videoInfo ? !!(bid.params.stream &&
+      bid.params.stream.video &&
+      bid.params.stream.video.mimes &&
+      bid.params.stream.video.mimes.length > 0 &&
+      bid.params.stream.protocols &&
+      bid.params.stream.protocols.length > 0) : true));
+
+  if (!isValid) {
+    utils.logError('Validation Error');
+  }
+
+  return isValid;
+}
+
+function hasValue(value) {
+  return (
+    value !== undefined &&
+    value !== null
+  );
+}
+
+function getGdprConsentChoice(bidderRequest) {
+  const hasGdprConsent =
+    hasValue(bidderRequest) &&
+    hasValue(bidderRequest.gdprConsent);
+
+  if (hasGdprConsent) {
+    return bidderRequest.gdprConsent;
+  }
+
+  return null;
+}
+
+export const spec = {
+  aliases: ['exads'], // short code
+  supportedMediaTypes: [BANNER, NATIVE, VIDEO],
+  isBidRequestValid: function (bid) {
+    utils.logInfo('on isBidRequestValid -> bid:', bid);
+
+    if (bid.bidder !== BIDDER) {
+      utils.logError('Validation Error', 'bidder wrong');
+      return false;
+    } else if (!bid.params.partner) {
+      utils.logError('Validation Error', 'bid.params.partner missed');
+      return false;
+    } else if (!Object.values(PARTNERS).includes(bid.params.partner)) {
+      utils.logError('Validation Error', 'bid.params.partner is not valid');
+      return false;
+    }
```

Comment:
Won't is be preferable to default to `RTB_2_4` if the partner value is missing? Especially given that we only support this format.

Reply:
In my opinion the correct partner is a **very critical parameter**, the most important of all after bidder code. It triggers one subset of code and if it is wrong or the publisher user doesn't know exactly how to use it, it could be a real problem.

It's true that currently we are supporting only RTB 2.4 but, **based on past comments**, we should support others partners in the next feature as exadsRTB or RTB 2.5 and/or others.

If we were sure that we won't support any other partner in the feature we could consider RTB the only partner and removing it as external parameter.